### ISSUE TITLE
[bug] YouTube tracks not playing in unsupported regions

### ISSUE THREAD

**Issue by alexmercerind** [harmonoid/harmonoid#155]
**Update By @mytja:**

You can now set proxy in settings.
Set it to `https://harmonoidproxy-eu.herokuapp.com/music`.
You should be able to use YouTube Music even if it isn't available in your country.
Any feedback is appreciated, just comment down below.

<hr></hr>

Few users have reported that they're unable to listen music from YouTube.

This is because YouTube Music isn't available in your country.
From next release we will seek some solution to this or switch to normal YouTube. 💜

Thanks for checking out the project. You can "subscribe" to this issue from the button on the right if you wish to get updates about this.

**Comment by FireMasterK** [comment]
Hey, Author of [Piped](https://github.com/TeamPiped/Piped) here, Piped is a privacy-friendly alternative frontend for YouTube, I have a public API that can be accessed.

Piped's backend uses [NewPipeExtractor](https://github.com/TeamNewPipe/NewPipeExtractor), a library that's used to extract data from YouTube, which implements all functionality present in https://github.com/harmonoid/harmonoid/blob/master/lib/youtube/youtube_api.dart with the exception of the `getRecommendations` function, since only YouTube is supported.

I have automatically generated OpenAPI bindings at https://github.com/TeamPiped/piped_dart, however, this might be overkill for this project.

I've heard from a couple of users that Piped works even in countries where YouTube is blocked, maybe perhaps you could include Piped as a fallback option.

**Comment by alexmercerind** [comment]
Hey @FireMasterK !

We had a look at your API & Dart bindings.

Your project looks really awesome. We will decide more in the future. However, for now we have decided to keep using our own little class for making the web requests for now since adding another proxy layer in between will increase the latency & introduce more delay in the playback of the tracks. I'm bit sceptical about depending on some third-party server for getting the content & YouTube isn't technically our primary point of focus at this point.


Thanks for introducing us with your open bindings here! Keep up the good work.

**Comment by talhakhan1297** [comment]
I downloaded this app just for the YouTube feature and it is not supported in my country. But still the app is great. Looking forward to the solution of this issue.

**Comment by mytja** [comment]
Hello,
so, here we have some updates on this critical issue. We've created our own proxy server for YouTube & YouTube Music, called `HarmonoidProxy`. I've deployed it (as it is currently private repository - hopefully soon getting public) to Heroku. Once PR #251 gets merged, you'll be able to download nightly release and test it out yourself.
Thank you so much

**Comment by mytja** [comment]
You can try [latest nightly release](https://github.com/harmonoid/harmonoid-nightly/releases/tag/475ae0c)
You can set proxy in settings. Set it to `https://harmonoidproxy-eu.herokuapp.com/music` and you should be able to use YouTube Music even if it isn't available in your country. Any feedback is appreciated, just comment down below.

**Comment by mytja** [comment]
Just a little update. We've released v0.2.2, which contains the proxy options in settings.
Also, we made [or YouTube Music proxy](https://github.com/harmonoid/ytm-proxy) public and thus available for anyone to deploy. You can deploy it to Heroku or your own virtual machine.

**Comment by thangisme** [comment]
> This has been merged into `master`. You can try [latest nightly release](https://github.com/harmonoid/harmonoid-nightly/releases/tag/475ae0c) You can set proxy in settings. Set it to `https://harmonoidproxy-eu.herokuapp.com/music` and you should be able to use YouTube Music even if it isn't available in your country. Any feedback is appreciated, just comment down below.
Hi, I followed your instruction and deployed ytmusic proxy server and put the url in the settings. Afterwards, the app can retrieve music suggestion, image cover, lyrics,... But It can't load the music itself. I have to enable VPN to make it work. Is there anything wrong?

**Comment by alexmercerind** [comment]
> > This has been merged into `master`. You can try [latest nightly release](https://github.com/harmonoid/harmonoid-nightly/releases/tag/475ae0c) You can set proxy in settings. Set it to `https://harmonoidproxy-eu.herokuapp.com/music` and you should be able to use YouTube Music even if it isn't available in your country. Any feedback is appreciated, just comment down below.
>
> Hi, I followed your instruction and deployed ytmusic proxy server and put the url in the settings. Afterwards, the app can retrieve music suggestion, image cover, lyrics,... But It can't load the music itself. I have to enable VPN to make it work. Is there anything wrong?

Hi @thangisme, are you able to search from the "YouTube Music" section? If so, are you able to play music videos (instead of music track) ?

Thanks! Your help will be crucial for making things work where YouTube Music isn't available.

**Comment by mytja** [comment]
>> This has been merged into master. You can try [latest nightly release](https://github.com/harmonoid/harmonoid-nightly/releases/tag/475ae0c) You can set proxy in settings. Set it to https://harmonoidproxy-eu.herokuapp.com/music and you should be able to use YouTube Music even if it isn't available in your country. Any feedback is appreciated, just comment down below.
>
> Hi, I followed your instruction and deployed ytmusic proxy server and put the url in the settings. Afterwards, the app can retrieve music suggestion, image cover, lyrics,... But It can't load the music itself. I have to enable VPN to make it work. Is there anything wrong?

This is a known issue for quite a bit, at least for me. What we're doing is basically using an [internal proxy](https://github.com/harmonoid/ytm-proxy) (not just any public (standardized) proxy), but that proxy is never receiving requests to the `/youtubei/v1/player` endpoint, as you can see from the logs.

Why, you might ask? We do all the searching on the [ytm_client.dart](https://github.com/harmonoid/ytm_client.dart) library/package, which ***CAN*** use the [ytm-proxy](https://github.com/harmonoid/ytm-proxy).

What's going on is, that we don't request the streams (audio files) from YouTube through the [ytm_client.dart](https://github.com/harmonoid/ytm_client.dart) library/package, but instead we pass the URL of the YouTube (Music) video to [mpv](https://mpv.io) and our [libmpv.dart](https://github.com/harmonoid/libmpv.dart) mpv Dart wrapper, which ***DOES NOT*** use the [ytm-proxy](https://github.com/harmonoid/ytm-proxy). [mpv](https://mpv.io) does all the stream retrieving requests for us through [youtube-dl](https://github.com/ytdl-org/youtube-dl), which ***DOES NOT*** know anything about our proxy (doesn't even know, we're using a proxy) and thus makes requests without a proxy.

Why do the thumbnails work, you might ask, as we also don't fetch them with our proxy?
1. YouTube serves thumbnails from a different URL than "youtube.com". If your government has blocked YouTube (and all its other services) it might be possible, that that domain isn't blocked.
2. If YouTube Music isn't yet available in your country, this doesn't matter - YouTube uses the same domain for serving the thumbnails no matter what service you're using. Infact, all music videos on YouTube Music are just plain automatically generated videos on YouTube - try removing `music.` from your URL (domain), and make it `youtube.com`, you'll see that every video on YouTube Music is a video on YouTube.

Why does the lyrics server work?
- We are not using the [ytm-proxy](https://github.com/harmonoid/ytm-proxy) for the lyrics server.
- What we're essentially doing, is basically requesting the lyrics for every song to Spotify (Musixmatch). They give us the lyrics, and we pass them back to you. You (your Harmonoid client) have/has 0 interaction with Spotify, only with our servers, which are not blocked by any government. Infact, our lyric server is acting similarly as a proxy to Spotify, except we add our secret credentials needed for the lyrics to fetch.

I hope this clarifies this topic

**Comment by alexmercerind** [comment]
> > > This has been merged into master. You can try [latest nightly release](https://github.com/harmonoid/harmonoid-nightly/releases/tag/475ae0c) You can set proxy in settings. Set it to https://harmonoidproxy-eu.herokuapp.com/music and you should be able to use YouTube Music even if it isn't available in your country. Any feedback is appreciated, just comment down below.
> > 
> > 
> > Hi, I followed your instruction and deployed ytmusic proxy server and put the url in the settings. Afterwards, the app can retrieve music suggestion, image cover, lyrics,... But It can't load the music itself. I have to enable VPN to make it work. Is there anything wrong?
> 
> This is a known issue for quite a bit, at least for me. What we're doing is basically using an [internal proxy](https://github.com/harmonoid/ytm-proxy) (not just any public (standardized) proxy), but that proxy is never receiving requests to the `/youtubei/v1/player` endpoint, as you can see from the logs.
> 
> Why, you might ask? We do all the searching on the [ytm_client.dart](https://github.com/harmonoid/ytm_client.dart) library/package, which _**CAN**_ use the [ytm-proxy](https://github.com/harmonoid/ytm-proxy).
> 
> What's going on is, that we don't request the streams (audio files) from YouTube through the [ytm_client.dart](https://github.com/harmonoid/ytm_client.dart) library/package, but instead we pass the URL of the YouTube (Music) video to [mpv](https://mpv.io) and our [libmpv.dart](https://github.com/harmonoid/libmpv.dart) mpv Dart wrapper, which _**DOES NOT**_ use the [ytm-proxy](https://github.com/harmonoid/ytm-proxy). [mpv](https://mpv.io) does all the stream retrieving requests for us through [youtube-dl](https://github.com/ytdl-org/youtube-dl), which _**DOES NOT**_ know anything about our proxy (doesn't even know, we're using a proxy) and thus makes requests without a proxy.
> 
> Why do the thumbnails work, you might ask, as we also don't fetch them with our proxy?
> 
> 1. YouTube serves thumbnails from a different URL than "youtube.com". If your government has blocked YouTube (and all its other services) it might be possible, that that domain isn't blocked.
> 2. If YouTube Music isn't yet available in your country, this doesn't matter - YouTube uses the same domain for serving the thumbnails no matter what service you're using. Infact, all music videos on YouTube Music are just plain automatically generated videos on YouTube - try removing `music.` from your URL (domain), and make it `youtube.com`, you'll see that every video on YouTube Music is a video on YouTube.
> 
> Why does the lyrics server work?
> 
> * We are not using the [ytm-proxy](https://github.com/harmonoid/ytm-proxy) for the lyrics server.
> * What we're essentially doing, is basically requesting the lyrics for every song to Spotify (Musixmatch). They give us the lyrics, and we pass them back to you. You (your Harmonoid client) have/has 0 interaction with Spotify, only with our servers, which are not blocked by any government. Infact, our lyric server is acting similarly as a proxy to Spotify, except we add our secret credentials needed for the lyrics to fetch.
> 
> I hope this clarifies this topic

Our aim is to solve the issue by listening to the things reported by the user.
I don't think so much redundant clarification is needed by a common user (which STILL doesn't result in any solution).
And, no youtube-dl isn't used either. I have other inference from the comment.

**Comment by thangisme** [comment]
> > > This has been merged into `master`. You can try [latest nightly release](https://github.com/harmonoid/harmonoid-nightly/releases/tag/475ae0c) You can set proxy in settings. Set it to `https://harmonoidproxy-eu.herokuapp.com/music` and you should be able to use YouTube Music even if it isn't available in your country. Any feedback is appreciated, just comment down below.
> > 
> > 
> > Hi, I followed your instruction and deployed ytmusic proxy server and put the url in the settings. Afterwards, the app can retrieve music suggestion, image cover, lyrics,... But It can't load the music itself. I have to enable VPN to make it work. Is there anything wrong?
> 
> Hi @thangisme, are you able to search from the "YouTube Music" section? If so, are you able to play music videos (instead of music track) ?
> 
> Thanks! Your help will be crucial for making things work where YouTube Music isn't available.

Yes, I'm able to search for song and play music video, but not the music track. I though the proxy server proxy all the request?

**Comment by alexmercerind** [comment]
No longer in plans.
